modules/llrt_crypto/src/subtle/mod.rs

```diff
@@ -0,0 +1,404 @@
+    match curve.to_ascii_uppercase().as_str() {
+        "P-256" => Ok(CryptoNamedCurve::P256),
+        "P-384" => Ok(CryptoNamedCurve::P384),
+        _ => Err(Exception::throw_message(ctx, "named_curve not found")),
+    }
+}
+
+#[derive(Debug)]
+pub enum DeriveAlgorithm {
+    Edch {
+        curve: CryptoNamedCurve,
+        public: Vec<u8>,
+    },
+    Hkdf {
+        hash: Sha,
+        salt: Vec<u8>,
+        info: Vec<u8>,
+    },
+    Pbkdf2 {
+        hash: Sha,
+        salt: Vec<u8>,
+        iterations: u32,
+    },
+}
+
+#[derive(Debug)]
+pub enum KeyGenAlgorithm {
+    Rsa {
+        modulus_length: u32,
+        public_exponent: Vec<u8>,
+    },
+    Ec {
+        curve: CryptoNamedCurve,
+    },
+    Aes {
+        length: u32,
+    },
+    Hmac {
+        hash: Sha,
+        length: Option<u32>,
+    },
+}
+
+pub async fn subtle_decrypt<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    key: ObjectBytes<'js>,
+    data: ObjectBytes<'js>,
+) -> Result<ArrayBuffer<'js>> {
+    let algorithm = extract_algorithm_object(&ctx, &algorithm)?;
+
+    let bytes = decrypt(&ctx, &algorithm, key.as_bytes(), data.as_bytes())?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
+}
+
+pub async fn subtle_derive_bits<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    base_key: ObjectBytes<'js>,
+    length: u32,
+) -> Result<ArrayBuffer<'js>> {
+    let derive_algorithm = extract_derive_algorithm(&ctx, &algorithm)?;
+
+    let bytes = derive_bits(&ctx, &derive_algorithm, base_key.as_bytes(), length)?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
+}
+
+pub async fn subtle_digest<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    data: ObjectBytes<'js>,
+) -> Result<ArrayBuffer<'js>> {
+    let algorithm = if let Some(algorithm) = algorithm.as_string() {
+        algorithm.to_string().or_throw(&ctx)?
+    } else {
+        algorithm
+            .get_optional::<_, String>("name")?
+            .ok_or_else(|| {
+                Exception::throw_message(&ctx, "Missing algorithm name should cause TypeError")
+            })?
+    };
+
+    let bytes = digest(&ctx, &algorithm, data.as_bytes())?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
+}
+
+pub async fn subtle_encrypt<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    key: ObjectBytes<'js>,
+    data: ObjectBytes<'js>,
+) -> Result<ArrayBuffer<'js>> {
+    let algorithm = extract_algorithm_object(&ctx, &algorithm)?;
+
+    let bytes = encrypt(&ctx, &algorithm, key.as_bytes(), data.as_bytes())?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
+}
+
+pub async fn subtle_generate_key<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    _extractable: bool,
+    _key_usages: Array<'js>,
+) -> Result<ArrayBuffer<'js>> {
+    let key_gen_algorithm = extract_generate_key_algorithm(&ctx, &algorithm)?;
+
+    let bytes = generate_key(&ctx, &key_gen_algorithm)?;
+    ArrayBuffer::new(ctx, bytes.as_slice())
```

Comment:
Do not pass slice as this will be an extra copy. Pass the vec directly to `ArrayBuffer`